openssl/aes.go

```diff
@@ -423,34 +352,49 @@ func (g *aesGCM) Seal(dst, nonce, plaintext, additionalData []byte) []byte {
 	if len(dst)+len(plaintext)+gcmTagSize < len(dst) {
 		panic("cipher: message too large for buffer")
 	}
+	if g.tls {
+		if len(additionalData) != gcmTlsAddSize {
+			panic("cipher: incorrect additional data length given to GCM TLS")
+		}
+		// BoringCrypto enforces strictly monotonically increasing explicit nonces
+		// and to fail after 2^64 - 1 keys as per FIPS 140-2 IG A.5,
+		// but OpenSSL does not perform this check, so it is implemented here.
```

Comment:
We added this check to honor `crypto/tls` expectation that `boring.NewGCMTLS` validates key/iv pair uniqueness. If this is not implemented, then `boring.NewGCMTLS` and `aesCipher.NewGCM` will behave the same. Yet, no test is failing if this check is not implemented, as upstream is not testing this property.

FIPS 140-3 relaxed key-iv pair uniqueness constraints from SP 800-38D in some cases and added a couple of more options to be in compliance with this requirement, but I don't see where it downgrades this requirement to a recommendation. I reluctant to remove this check, as SP 800-38D Sec. 8 states that

> Compliance with this requirement is crucial to the security of GCM.

IMO doing nothing is not an option (I might be wrong here, correct me if that's the case), we either check it ourselves or force OpenSSL to do so, and the latter is not straightforward due to API incompatibilities.

About the lowered key usage limit in TLS 1.3, we could implement it in a latter PR so this current PR does not contain new features not available in either RedHat or Microsoft implementations.

Reply:
If the logic is aligned with `crypto/tls` then let's have it as it is; my concern was mainly that it might blur the boundary of FIPS module as the check shall be performed within a module.